aten/src/ATen/native/native_functions.yaml

```diff
@@ -5821,6 +5823,8 @@
 - func: masked_fill.Scalar(Tensor self, Tensor mask, Scalar value) -> Tensor
   device_check: NoCheck   # TensorIterator
   variants: function, method
+  dispatch:
+    CPU, CUDA: masked_fill
```

Comment:
Is there a reason why these can't be `CompositeExplicitAutograd` instead?

I think that's why the XLA tests are failing in https://github.com/pytorch/xla/issues/3314: `masked_fill` is no longer composite, so calling `at::masked_fill()` with an xla tensor will dispatch directly to the `XLA` key. There's no kernel registered directly to `masked_fill()`, so we hit the CPU fallback, causing the test to fail.

If it were `CompositeExplicitAutograd` instead, then XLA would end up re-using the `masked_fill` kernel in core, which redispatches to `masked_fill_()` (which XLA has a lowering for).

Reply:
My understanding is that CompositeExplicitAutograd operations should work for all/most backends (please correct me if that's not supposed to be the case).

I didn't make this CompositeExplicitAutograd because its implementation is `self.clone().masked_fill_(source)`.  masked_fill_ only has CPU, CUDA dispatch, so masked_fill only should support the CPU and CUDA backends.

That being said I don't know how XLA works and am happy to change this to CompositeExplicitAutograd if we think it's the right thing to do.

Reply:
You're right that since the implementation is `self.clone().masked_fill_(source)`, then even though it's "composite", a backend will only support it if it provides kernels for those ops (`clone` and `masked_fill_`).

In this case though, that's pretty much what the current state is today, since `masked_fill()` is already a `CompositeImplicitAutograd` op in master. And XLA [currently has an explicit lowering](https://github.com/pytorch/xla/blob/343b475ffd2a3c505531dd93a05a4cdf924c0ed3/xla_native_functions.yaml#L176) for `masked_fill_()`, basically for that reason.

So I think that we should make these kernels all `CompositeExplicitAutograd` for now to preserve the current state. The better long term state will be for these ops to all become structured (and XLA will be able to just directly lower the out-of-place `masked_fill()` instead.

Reply:
I am not sure that is the idea.
If a backend wants to override this op, then if we specify the keys for both, they have to re-implement both functions.
But if masked_fill is compositeexplicit. Then they only have to implement masked_fill_ to get both working.
So I would agree with @bdhirsh here.

Reply:
Cool, thanks for the discussion, I will change to CompositeExplicitAutograd.

Reply:
Question: does changing it to CompositeExplicitAutograd prevent XLA from registering a lowering for e.g. masked_fill if they desired? For a in-tree backend I expect this would be possible by specifying the different dispatch for different backend.

Reply:
Nope - Registering a kernel as `CompositeExplicitAutograd` means that every backend will get that kernel by default, but can override with their own kernel if they want.

XLA also has a boxed fallback for all ops that falls back to CPU. But that gets an even lower precedence than the composite kernel. The full precedence order defined in the dispatcher registration logic is:
(1) direct kernel registrations
(2) composite registrations (if no direct registration defined)
(3) boxed fallback (if 1 and 2 both aren't available)

Reply:
Awesome, thanks for the detailed explanation!